benchmarks/float8/bench_grouped_mm.py

```diff
@@ -0,0 +1,222 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD 3-Clause license found in the
+# LICENSE file in the root directory of this source tree.
+from typing import Optional
+
+import fire
+import pandas as pd
+import torch
+import torch.utils.benchmark as benchmark
+from utils import (
+    get_gpu_kernel_gemm_time_s,
+)
+
+from torchao.testing.training.roofline_utils import get_specs
+
+
+def benchmark_fn_in_sec(f, *args, **kwargs):
+    # Manual warmup
+    for _ in range(4):
+        f(*args, **kwargs)
+    t0 = benchmark.Timer(
+        stmt="f(*args, **kwargs)", globals={"args": args, "kwargs": kwargs, "f": f}
+    )
+    measurement = t0.blocked_autorange()
+    return measurement.mean
+
+
+def do_benchmarks(
+    tops,
+    peak_tops,
+    use_gpu_kernel_time,
+    f,
+    *args,
+    **kwargs,
+):
+    if use_gpu_kernel_time:
+        # just the gemm GPU kernel
+        time_sec = get_gpu_kernel_gemm_time_s(f, *args, **kwargs)
+    else:
+        # e2e time including kernel launch overhead
+        time_sec = benchmark_fn_in_sec(f, *args, **kwargs)
+    tops_sec = float(tops) / time_sec
+    pct_top_peak = tops_sec / peak_tops
+    return time_sec, tops_sec, pct_top_peak
+
+
+@torch.inference_mode()
+def run(
+    n_limit: Optional[int] = None,
+    out_filename: Optional[str] = None,
+    M: Optional[int] = None,
+    K: Optional[int] = None,
+    N: Optional[int] = None,
+    E: Optional[int] = None,  # dim 0 of B tensor (num experts)
+    use_gpu_kernel_time: bool = True,
+    shape_gen_name="llama4_17bx16e",
+    recipe: str = "rowwise",
+):
+    device = "cuda"
+
+    assert recipe in ("rowwise",), "unsupported"
+
+    specs = get_specs()
+    bf16_peak_tops = specs["bf16_peak_tops"]
+    fp8_peak_tops = specs["fp8_peak_tops"]
+    print(f"gpu_name: {torch.cuda.get_device_name(0)}")
+    print(f"peak tops: bf16 {bf16_peak_tops:.2e}, fp8 {fp8_peak_tops:.2e}")
+    headers = (
+        "name",
+        "M",
+        "K",
+        "N",
+        "E",
+        "time_s",
+        "speedup",
+        "fp8_speedup",
+    )
+    results = []
+
+    dtype = torch.bfloat16
+    name_to_shapes = get_name_to_shapes_iter(shape_gen_name, M, K, N, E)
+
+    for idx, (name, (M, K, N, E)) in enumerate(
+        name_to_shapes,
+    ):
+        if n_limit is not None and idx >= n_limit:
+            break
+        assert M % E == 0, (
+            "tokens (M) must be evenly divisible by num experts (E) for this benchmark"
+        )
+        tops = 2 * M * N * K * E
+        print("M, K, N, E:", M, K, N, E, f"tops: {tops:.2E}")
+
+        # Run bf16 torch._grouped_mm baseline.
+        A = torch.randn(M, K, device=device, dtype=dtype)
+        B = torch.randn(E, K, N, device=device, dtype=dtype)
+        group_size = M // E
+        offs = torch.arange(
```

Comment:
should we add a way to control how jagged the individual gemms are? I'd guess perfectly balanced would get the largest and more consistent speedup, but a less perfect distribution is more representative of real use cases.

Reply:
Good idea, added `generate_jagged_offs` function and use it instead of these uniformly sized groups, which are not realistic at all.